### Задание 1. Базовое изучение  
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [5]:
import pandas as pd
import numpy as np

In [35]:
df = pd.read_csv('C:/Netology/Stat_Basic/horse_data.csv', names = ['surgery', 'age', 'Hospital Number', 'rectal temperature', 
'pulse', 'respiratory rate', 'temperature of extremities','peripheral pulse', 'mucous membranes', 'capillary refill time', 
'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube','nasogastric reflux', 'nasogastric reflux PH',
'rectal examination - feces', 'abdomen', 'packed cell volume', 'total_protein','abdominocentesis appearance', 
'abdomcentesis total protein', 'outcome', 'first number is site of lesion','second number is type', 
'third number is subtype', 'fourth number is specific code', 'cp_data'], na_values = '?')

horse_info = df[['surgery', 'age', 'pulse', 'temperature of extremities', 'peripheral pulse', 'pain', 'total_protein', 'outcome']]
horse_info.head(10)

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         NaN               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         NaN               NaN   NaN   
5      2.0    1    NaN                         2.0               1.0   2.0   
6      1.0    1   48.0                         1.0               1.0   3.0   
7      1.0    1   60.0                         3.0               NaN   NaN   
8      2.0    1   80.0                         3.0               4.0   4.0   
9      2.0    9   90.0                         1.0               NaN   5.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0  
5            NaN      1.0  
6            7.0      1.0  
7            8.3      2.0  
8            6.2      3.0  
9            6.2      1.0

In [8]:
horse_info.describe()

surgery         age       pulse  temperature of extremities  \
count  299.000000  300.000000  276.000000                  244.000000   
mean     1.397993    1.640000   71.913043                    2.348361   
std      0.490305    2.173972   28.630557                    1.045054   
min      1.000000    1.000000   30.000000                    1.000000   
25%      1.000000    1.000000   48.000000                    1.000000   
50%      1.000000    1.000000   64.000000                    3.000000   
75%      2.000000    1.000000   88.000000                    3.000000   
max      2.000000    9.000000  184.000000                    4.000000   

       peripheral pulse       pain  total_protein     outcome  
count        231.000000  245.00000     267.000000  299.000000  
mean           2.017316    2.95102      24.456929    1.551839  
std            1.042428    1.30794      27.475009    0.737187  
min            1.000000    1.00000       3.300000    1.000000  
25%            1.000000    2.00000       6.500000    1.000000  
50%            2.000000    3.00000       7.500000    1.000000  
75%            3.000000    4.00000      57.000000    2.000000  
max            4.000000    5.00000      89.000000    3.000000

In [9]:
moda = horse_info.round().mode()
moda

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      1.0    1   48.0                         3.0               1.0   3.0   

   total_protein  outcome  
0            7.0      1.0

In [10]:
horse_info.groupby('peripheral pulse').count()['age']

peripheral pulse
1.0    115
2.0      5
3.0    103
4.0      8
Name: age, dtype: int64

In [11]:
horse_info.groupby('outcome').count()['age']

outcome
1.0    178
2.0     77
3.0     44
Name: age, dtype: int64

1. В столбце "Age"  есть значение 9. Думаю ошибка в исходных данных.
2. Большая часть лошадей выжила.
3. У половины лошадей показатель "Total protein" многократно завышен.Средний разброс значений огромен по сравнению с нормальным диапазоном значений

### Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [12]:
q1 = horse_info.pulse.quantile(0.25)
q3 = horse_info.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)

outliers = horse_info[(horse_info.pulse > upper_bound) | (horse_info.pulse < lower_bound)].sort_values('pulse')
outliers

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
41       2.0    9  150.0                         NaN               NaN   NaN   
275      1.0    9  150.0                         1.0               3.0   5.0   
55       1.0    9  160.0                         3.0               NaN   3.0   
3        1.0    9  164.0                         4.0               1.0   2.0   
255      1.0    9  184.0                         1.0               NaN   4.0   

     total_protein  outcome  
41             8.5      1.0  
275            6.2      2.0  
55             NaN      2.0  
3              7.2      2.0  
255            3.3      2.0

In [19]:
q1 = horse_info.total_protein.quantile(0.25)
q3 = horse_info.total_protein.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
outliers = horse_info[(horse_info.total_protein < lower_bound) | (horse_info.total_protein > upper_bound)].sort_values('total_protein')
outliers

Empty DataFrame
Columns: [surgery, age, pulse, temperature of extremities, peripheral pulse, pain, total_protein, outcome]
Index: []

In [20]:
q1 = horse_info.age.quantile(0.25)
q3 = horse_info.age.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)

outliers = horse_info[(horse_info.age > upper_bound) | (horse_info.age < lower_bound)].sort_values('pulse')
outliers

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
287      2.0    9   84.0                         NaN               NaN   NaN   
109      1.0    9   84.0                         3.0               3.0   3.0   
262      2.0    9   88.0                         NaN               NaN   NaN   
9        2.0    9   90.0                         1.0               NaN   5.0   
13       2.0    9   92.0                         1.0               1.0   1.0   
75       1.0    9  100.0                         3.0               3.0   2.0   
82       1.0    9  100.0                         3.0               1.0   3.0   
191      2.0    9  100.0                         2.0               1.0   4.0   
229      1.0    9  120.0                         NaN               NaN   NaN   
212      1.0    9  120.0                         1.0               3.0   3.0   
252      1.0    9  124.0                         3.0               1.0   2.0   
244      1.0    9  124.0                         1.0               3.0   2.0   
16       1.0    9  128.0                         3.0               3.0   4.0   
23       1.0    9  130.0                         NaN               3.0   2.0   
135      1.0    9  136.0                         3.0               3.0   5.0   
103      1.0    9  140.0                         1.0               1.0   3.0   
39       1.0    9  146.0                         NaN               NaN   NaN   
41       2.0    9  150.0                         NaN               NaN   NaN   
275      1.0    9  150.0                         1.0               3.0   5.0   
55       1.0    9  160.0                         3.0               NaN   3.0   
3        1.0    9  164.0                         4.0               1.0   2.0   
255      1.0    9  184.0                         1.0               NaN   4.0   
74       1.0    9    NaN                         NaN               NaN   NaN   
227      1.0    9    NaN                         NaN               NaN   NaN   

     total_protein  outcome  
287            5.0      1.0  
109            6.6      2.0  
262            5.9      1.0  
9              6.2      1.0  
13             6.1      2.0  
75            57.0      3.0  
82             5.7      1.0  
191            4.7      1.0  
229           54.0      1.0  
212            6.3      1.0  
252            7.6      2.0  
244            8.0      1.0  
16             7.8      2.0  
23            70.0      1.0  
135            4.9      2.0  
103            5.3      1.0  
39             NaN      2.0  
41             8.5      1.0  
275            6.2      2.0  
55             NaN      2.0  
3              7.2      2.0  
255            3.3      2.0  
74             4.9      2.0  
227            6.5      1.0

1. Можно предположить, что значение 9 выставлено для молодых особей, т.к. пульс у молодых особей выше, чем у взрослых. Меняем 9 на 2.
2. Значения от 150 до 164 не считаю выбросами. Значение 184 попадает под сомнение. Тем не менее, не счиатаю его выбросом, т.к. у данной лошади наблюдается низкое значение протеина, сильные болевые ощущения. Вполне вероятно, что она находилась в критическом состоянии. Это подтверждает её смерть в итоге.
3. Выбросов по столбцу "total_protein" не выявлено.

### Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [22]:
for empty_values in horse_info.columns:
    quantity = horse_info[empty_values].isnull().sum() 
    percent = quantity / len(horse_info[empty_values])
 
    print(empty_values, ' - ', quantity,  '(', round(percent * 100, 1), '%)', 'пропусков')

surgery  -  1 ( 0.3 %) пропусков
age  -  0 ( 0.0 %) пропусков
pulse  -  24 ( 8.0 %) пропусков
temperature of extremities  -  56 ( 18.7 %) пропусков
peripheral pulse  -  69 ( 23.0 %) пропусков
pain  -  55 ( 18.3 %) пропусков
total_protein  -  33 ( 11.0 %) пропусков
outcome  -  1 ( 0.3 %) пропусков


- surgery : 1 пропуск - исключить
- pulse: 24 пропуска, 8% от общих данных. Заполнение: группировка по возрасту и показателю "Переферический пульс". Заполнение: медианой.
- temperature of extremities: 56 пропусков, 18,7%. Заполним медианным по surgery, age, outcome.
- peripheral pulse: 69 пропусков, что составляет 23,1% от общих данных, что существенно,заполним медианным по pulse.
- pain: так как признак является категориальным, заполним медианным по surgery, age, outcome
- total_protein:11% пропущено. Заполним недостающие данные медианой.
- outcome: 1 пропуск - исключить


In [52]:
df_rezult = horse_info

In [61]:
# surgery 
df_rezult = horse_info.dropna(subset=['surgery'])
df_rezult.head()

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [62]:
# pulse
df_rezult['pulse'].fillna(horse_info.groupby(['age', 'peripheral pulse'])['pulse'].transform('median'), inplace = True)
df_rezult.head()

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [63]:
# temperature of extremities
df_rezult['temperature of extremities'].fillna(horse_info.groupby(['surgery','age','outcome'])['temperature of extremities'].transform('median'), inplace=True)
df_rezult.head()

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [65]:
#pain
df_rezult['pain'].fillna(horse_info.groupby(['surgery','age','outcome'])['pain'].transform('median'), inplace=True)
df_rezult.head()

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [66]:
# outcome
df_rezult = horse_info.dropna(subset=['outcome'])
df_rezult.head()

surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [76]:
# peripheral pulse
df_rezult['peripheral pulse'].fillna(horse_info.groupby(['age', 'pulse'])['peripheral pulse'].transform('median'), inplace = True)
df_rezult = horse_info.dropna(subset=['peripheral pulse'])
df_rezult.head()

C:\Anaconda\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
5      2.0    1   50.0                         2.0               1.0   2.0   
6      1.0    1   48.0                         1.0               1.0   3.0   

   total_protein  outcome  
0            8.4      2.0  
2            6.7      1.0  
3            7.2      2.0  
5            NaN      1.0  
6            7.0      1.0

In [77]:
df_rezult['total_protein'].fillna(horse_info['total_protein'].mode()[0], inplace=True)
horse_info.head()

C:\Anaconda\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


surgery  age  pulse  temperature of extremities  peripheral pulse  pain  \
0      2.0    1   66.0                         3.0               3.0   5.0   
1      1.0    1   88.0                         3.0               NaN   3.0   
2      2.0    1   40.0                         1.0               1.0   3.0   
3      1.0    9  164.0                         4.0               1.0   2.0   
4      2.0    1  104.0                         3.0               NaN   4.0   

   total_protein  outcome  
0            8.4      2.0  
1           85.0      3.0  
2            6.7      1.0  
3            7.2      2.0  
4            7.4      2.0

In [78]:
df_rezult.isna().sum()

surgery                       0
age                           0
pulse                         0
temperature of extremities    0
peripheral pulse              0
pain                          0
total_protein                 0
outcome                       0
dtype: int64